lib/core/CDataFrame.cc

```diff
@@ -0,0 +1,513 @@
+        break;
+    }
+    return {m_NumberRows, std::move(m_SyncWrittenSlices)};
+}
+
+CDataFrame::TSizeSizePr CDataFrame::numberOfThreadsAndStride(std::size_t target) const {
+
+    if (m_Slices.empty()) {
+        return {1, 0};
+    }
+
+    std::size_t numberSlices{m_Slices.size()};
+    std::size_t numberThreads{std::min(numberSlices, target)};
+    std::size_t strideLowerBound{numberSlices / numberThreads};
+    std::size_t strideUpperBound{strideLowerBound + 1};
+
+    if (numberSlices % numberThreads == 0) {
+        return {numberThreads, strideLowerBound};
+    }
+    if (numberSlices % strideUpperBound == 0) {
+        return {numberThreads - 1, strideUpperBound};
+    }
+    return {numberThreads, strideUpperBound};
+}
+
+void CDataFrame::CDataFrameRowSliceWriter::finishAsyncWriteToStore() {
+    // Passing an empty slice signals to the thread writing slices to
+    // storage that we're done.
+
+    if (m_Writing) {
+        if (m_AsyncWriteToStore == EReadWriteToStorage::E_Async) {
+            m_SlicesToAsyncWriteToStore.push({0, TFloatVec{}});
+        }
+        m_Writing = false;
+    }
+}
+
+CDataFrame makeMainStorageDataFrame(std::size_t numberColumns,
+                                    boost::optional<std::size_t> sliceCapacity,
+                                    CDataFrame::EReadWriteToStorage readWriteToStoreAsync) {
+    // The return copy is elided so we never need to call the explicitly
+    // deleted data frame copy constructor.
+
+    auto writer = [](std::size_t firstRow, TFloatVec slice) {
+        return boost::make_unique<CMainMemoryDataFrameRowSlice>(firstRow, std::move(slice));
+    };
+
+    if (sliceCapacity != boost::none) {
+        return {numberColumns, *sliceCapacity, readWriteToStoreAsync, writer};
+    }
+
+    return {numberColumns, readWriteToStoreAsync, writer};
+}
+
+CDataFrame makeDiskStorageDataFrame(const std::string& rootDirectory,
+                                    std::size_t numberColumns,
+                                    std::size_t numberRows,
+                                    boost::optional<std::size_t> sliceCapacity,
+                                    CDataFrame::EReadWriteToStorage readWriteToStoreAsync) {
+    // The return copy is elided so we never need to call the explicitly
+    // deleted data frame copy constructor.
+
+    std::size_t minimumSpace{2 * numberRows * numberColumns * sizeof(CFloatStorage)};
+
+    COnDiskDataFrameRowSlice::TTemporaryDirectoryPtr directory{
+        std::make_shared<COnDiskDataFrameRowSlice::CTemporaryDirectory>(
+            rootDirectory, minimumSpace)};
+
+    // Note the writer lambda holds a reference to the directory shared
+    // pointer is copied to the data frame. So this isn't destroyed, and
+    // the folder cleaned up, until the data frame itself is destroyed.
+
+    auto writer = [directory](std::size_t firstRow, TFloatVec slice) {
+        return boost::make_unique<COnDiskDataFrameRowSlice>(directory, firstRow,
```

Comment:
If this is never going to be backported to 6.x you can use `std::make_unique`.  If it _is_ going to be backported to 6.x then there are potential problems with `std::async` not conforming to standard in VS2013: https://stackoverflow.com/questions/44829115/from-vs2013-to-vs2017-stdasync-does-not-start-a-new-thread

Reply:
I thought about this. I've used a couple of other C++14 features, for example capture with an initialiser so I can move variables into a lambda. However, at this point I suspect that we may want to back port this, so I wanted to keep non-portable features to a minimum and only where there was a real advantage.

I've been thinking in general it would be good to do a few more boost to std tidy ups in the code, particularly when we move to C++17 on the master. What I might do is raise an issue to cover this, but leave this as is now.